kcoin/genesis/genesis_generator.go

```diff
@@ -0,0 +1,404 @@
+	maxNumValidators              *big.Int
+	unbondingPeriod               *big.Int
+	walletAddressGenesisValidator *common.Address
+	prefundedAccounts             []*validPrefundedAccount
+	consensusEngine               string
+	smartContractsOwner           *common.Address
+}
+
+func GenerateGenesis(options GenesisOptions) (*core.Genesis, error) {
+	validOptions, err := validateOptions(options)
+	if err != nil {
+		return nil, err
+	}
+
+	genesis := &core.Genesis{
+		Timestamp: uint64(time.Now().Unix()),
+		GasLimit:  4700000,
+		Alloc:     make(core.GenesisAlloc),
+		Config:    &params.ChainConfig{},
+	}
+
+	setNetwork(validOptions.network, genesis)
+	setConsensusEngine(validOptions.consensusEngine, genesis)
+	setExtraData(options.ExtraData, genesis)
+
+	genesis.Alloc[*validOptions.smartContractsOwner] = core.GenesisAccount{Balance: new(big.Int).Mul(common.Big1, big.NewInt(params.Ether))}
+
+	//TODO: This maybe will be need to be available to change by the parameters in the options in the future, right now is 0.
+	baseDeposit := common.Big0
+
+	electionABI, err := abi.JSON(strings.NewReader(contracts.ElectionContractABI))
+	if err != nil {
+		return nil, err
+	}
+
+	electionParams, err := electionABI.Pack(
+		"",
+		baseDeposit,
+		validOptions.maxNumValidators,
+		validOptions.unbondingPeriod,
+		*validOptions.walletAddressGenesisValidator,
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	runtimeCfg := &runtime.Config{
+		Origin: *validOptions.smartContractsOwner,
+		EVMConfig: vm.Config{
+			Debug:  true,
+			Tracer: newVmTracer(),
+		},
+	}
+
+	contract, err := createContract(runtimeCfg, append(common.FromHex(contracts.ElectionContractBin), electionParams...))
+	if err != nil {
+		return nil, err
+	}
+
+	genesis.Alloc[contract.addr] = core.GenesisAccount{
+		Code:    contract.code,
+		Storage: contract.storage,
+		Balance: new(big.Int).Mul(baseDeposit, new(big.Int).SetUint64(params.Ether)),
+	}
+
+	setPrefundedAccounts(validOptions.prefundedAccounts, genesis)
+	addBatchOfPrefundedAccounts(genesis)
+
+	return genesis, nil
+}
+
+func setExtraData(extraData string, genesis *core.Genesis) {
+	extra := ""
+	if extraData != "" {
+		extra = extraData
+	}
+	genesis.ExtraData = make([]byte, 32)
+	if len(extra) > 32 {
+		extra = extra[:32]
+	}
+	genesis.ExtraData = append([]byte(extra), genesis.ExtraData[len(extra):]...)
+}
+
+func setConsensusEngine(consensusEngine string, genesis *core.Genesis) {
+	switch consensusEngine {
+	case TendermintConsensus:
+		genesis.Config.Tendermint = &params.TendermintConfig{Rewarded: true}
+	}
+}
+
+func setNetwork(network string, genesis *core.Genesis) {
+	switch network {
+	case MainNetwork:
+		genesis.Config.ChainID = params.MainnetChainConfig.ChainID
+	case TestNetwork:
+		genesis.Config.ChainID = params.TestnetChainConfig.ChainID
+	case OtherNetwork:
+		genesis.Config.ChainID = new(big.Int).SetUint64(uint64(rand.Intn(65536)))
+	}
+}
+
+func validateOptions(options GenesisOptions) (*validGenesisOptions, error) {
```

Comment:
umm, I have my doubts in this one, since finally for me it validates if the options are correct, and if they are they return a validOptions struct. Can it be?

Reply:
:) nw

does both things, depends on what you are seeing as main, does validate and map to a diff struct

You can leave as is